apps/cowswap-frontend/src/modules/permit/utils/generatePermitHook.ts

```diff
@@ -105,11 +107,43 @@ async function generatePermitHookRaw(params: PermitHookParams): Promise<PermitHo
 
   const callData = await callDataPromise
 
+  const gasLimit = await calculateGasLimit(callData, owner, tokenAddress, provider, !!account)
+
   return {
     target: tokenAddress,
     callData,
-    // TODO: Find a way to calculate the gas for quote calls without requesting user to sign
-    // For now, keep it hard coded
-    gasLimit: '80000',
+    gasLimit,
+  }
+}
+
+async function calculateGasLimit(
+  data: string,
+  from: string,
+  to: string,
+  provider: Web3Provider,
+  isUserAccount: boolean
+): Promise<string> {
+  if (isUserAccount) {
+    // "Real" user account, use the default
+    // If we send a smaller gas limit than what was used in the quote, it won't matter
+    // If we send a bigger gas limit, the order will fail
+
+    // TODO: it might still happen that user's permit will cost more than the static account
+    return DEFAULT_PERMIT_GAS_LIMIT
```

Comment:
Note that hooks have a hard gas limit on-chain. So if this is too small, then the hook will revert with `OUT_OF_GAS` and not work as intended (the API will likely reject the order with an `INSUFFICIENT_APPROVAL` error)

Reply:
Hmm I see.
I have not faced this yet while testing. I assume the default min of 80k has been enough so far.
Will then just re-use the same value that was calculated and used for the quote.

Reply:
Actually, a better approach might be to estimate also the real account, but do not add 10%.
It would fail anyway if the estimated amount in the first part is not enough for the second so, this is more accurate.

Reply:
It also has the added protection that you simulate the user `permit`, to catch any issues (like signature problems).

Reply:
Done